include/nbl/builtin/hlsl/bxdf/ndf/microfacet_to_light_transform.hlsl

```diff
@@ -0,0 +1,207 @@
+struct SDualMeasureQuant<T, false, MTT_REFLECT>
+{
+    using this_t = SDualMeasureQuant<T, false, MTT_REFLECT>;
+    using scalar_type = T;
+
+    NBL_CONSTEXPR_STATIC_INLINE MicrofacetTransformTypes Type = MTT_REFLECT;
+
+    scalar_type getMicrofacetMeasure()
+    {
+        return pdf;
+    }
+
+    // this computes the max(NdotL,0)/(4*max(NdotV,0)*max(NdotL,0)) factor which transforms PDFs in the f in projected microfacet f * NdotH measure to projected light measure f * NdotL
+    scalar_type getProjectedLightMeasure()
+    {
+        return scalar_type(0.25) * pdf / maxNdotV;
+    }
+
+    scalar_type pdf;
+    scalar_type maxNdotV;
+};
+
+template<typename T>
+struct SDualMeasureQuant<T, true, MTT_REFLECT>
+{
+    using this_t = SDualMeasureQuant<T, true, MTT_REFLECT>;
+    using scalar_type = T;
+
+    NBL_CONSTEXPR_STATIC_INLINE MicrofacetTransformTypes Type = MTT_REFLECT;
+
+    scalar_type getMicrofacetMeasure()
+    {
+        return pdf;
+    }
+
+    // this computes the max(NdotL,0)/(4*max(NdotV,0)*max(NdotL,0)) factor which transforms PDFs in the f in projected microfacet f * NdotH measure to projected light measure f * NdotL
+    scalar_type getProjectedLightMeasure()
+    {
+        return pdf * maxNdotL;
+    }
+
+    scalar_type pdf;
+    scalar_type maxNdotL;
+};
+
+template<typename T>
+struct SDualMeasureQuant<T, false, MTT_REFRACT>
+{
+    using this_t = SDualMeasureQuant<T, false, MTT_REFRACT>;
+    using scalar_type = T;
+
+    NBL_CONSTEXPR_STATIC_INLINE MicrofacetTransformTypes Type = MTT_REFRACT;
+
+    scalar_type getMicrofacetMeasure()
+    {
+        return pdf;
+    }
+
+    scalar_type getProjectedLightMeasure()
+    {
+        const scalar_type VdotH_etaLdotH = (VdotH + orientedEta * LdotH);
+        // VdotHLdotH is negative under transmission, so thats denominator is negative
+        scalar_type denominator = absNdotV * (-VdotH_etaLdotH * VdotH_etaLdotH);
+        return pdf * VdotHLdotH / denominator;
+    }
+
+    scalar_type pdf;
+    scalar_type absNdotV;
+    scalar_type VdotH;
+    scalar_type LdotH;
+    scalar_type VdotHLdotH;
+    scalar_type orientedEta;
+};
+
+template<typename T>
+struct SDualMeasureQuant<T, true, MTT_REFRACT>
+{
+    using this_t = SDualMeasureQuant<T, true, MTT_REFRACT>;
+    using scalar_type = T;
+
+    NBL_CONSTEXPR_STATIC_INLINE MicrofacetTransformTypes Type = MTT_REFRACT;
+
+    scalar_type getMicrofacetMeasure()
+    {
+        return pdf;
+    }
+
+    scalar_type getProjectedLightMeasure()
+    {
+        const scalar_type VdotH_etaLdotH = (VdotH + orientedEta * LdotH);
+        // VdotHLdotH is negative under transmission, so thats denominator is negative
+        scalar_type denominator = absNdotL * (-scalar_type(4.0) * VdotHLdotH / (VdotH_etaLdotH * VdotH_etaLdotH));
+        return pdf * denominator;
+    }
+
+    scalar_type pdf;
+    scalar_type absNdotL;
+    scalar_type VdotH;
+    scalar_type LdotH;
+    scalar_type VdotHLdotH;
+    scalar_type orientedEta;
```

Comment:
either take both `VdotHLdotH` and `VdotH_etaLdotH2`  as members (meaning you don't take `orientedEta`, `VdotH` or `LdotH`) , or you only take `VdotH`, `LdotH` and `orientedEta`